src/Compilers/CSharp/Portable/Binder/Semantics/Operators/UnaryOperatorOverloadResolution.cs

```diff
@@ -78,31 +76,55 @@ public void UnaryOperatorOverloadResolution(UnaryOperatorKind kind, BoundExpress
+            int currentBestIndex;
             var candidates = result.Results;
-            for (int i = 1; i < candidates.Count; ++i)
+            for (currentBestIndex = 0; currentBestIndex < candidates.Count; currentBestIndex++)
+            {
+                if (candidates[currentBestIndex].Kind == OperatorAnalysisResultKind.Applicable)
+                {
+                    break;
+                }
+            }
+
+            for (int index = currentBestIndex + 1; index < candidates.Count; index++)
+            {
+                if (candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, operand, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    candidates[index] = candidates[index].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[currentBestIndex] = candidates[currentBestIndex].Worse();
+                    currentBestIndex = index;
+                }
+            }
+
+            for (int index = 0; index < candidates.Count; index++)
             {
-                if (candidates[i].Kind != OperatorAnalysisResultKind.Applicable)
+                if (index == currentBestIndex || candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
                 {
                     continue;
                 }
 
-                // Is this applicable operator better than every other applicable method?
-                for (int j = 0; j < i; ++j)
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, operand, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
                 {
-                    if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
-                    {
-                        continue;
-                    }
-                    var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, operand, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
-                    {
-                        candidates[j] = candidates[j].Worse();
-                    }
-                    else if (better == BetterResult.Right)
-                    {
-                        candidates[i] = candidates[i].Worse();
-                    }
+                    candidates[index] = candidates[index].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[currentBestIndex] = candidates[currentBestIndex].Worse();
+                    break;
+                }
+                else
+                {
+                    break;
```

Comment:
It is hard to see how the problem I theorize could arise for the binary operators could arise for unary operators, because I expect the "better operator" relation is transitive for unary operators. But I could be wrong. If we adjust the code for binary operators, I recommend we do so here as well.